pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -579,49 +588,241 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+			}
+			// Before logging that UnmountDevice succeeded and moving on,
+			// use mounter.PathIsDevice to check if the path is a device,
+			// if so use mounter.DeviceOpened to check if the device is in use anywhere
+			// else on the system. Retry if it returns true.
+			isDevicePath, devicePathErr := mounter.PathIsDevice(deviceToDetach.DevicePath)
+			var deviceOpened bool
+			var deviceOpenedErr error
+			if !isDevicePath && devicePathErr == nil {
+				// not a device path or path doesn't exist
+				//TODO: refer to #36092
+				glog.V(3).Infof("Not checking device path %s", deviceToDetach.DevicePath)
+				deviceOpened = false
+			} else {
+				deviceOpened, deviceOpenedErr = mounter.DeviceOpened(deviceToDetach.DevicePath)
+				if deviceOpenedErr != nil {
+					return deviceToDetach.GenerateErrorDetailed("UnmountDevice.DeviceOpened failed", deviceOpenedErr)
+				}
+			}
+			// The device is still in use elsewhere. Caller will log and retry.
+			if deviceOpened {
+				return deviceToDetach.GenerateErrorDetailed(
+					"UnmountDevice failed",
+					fmt.Errorf("the device is in use when it was no longer expected to be in use"))
+			}
+
+			glog.Infof(deviceToDetach.GenerateMsgDetailed("UnmountDevice succeeded", ""))
+
+			// Update actual state of world
+			markDeviceUnmountedErr := actualStateOfWorld.MarkDeviceAsUnmounted(
+				deviceToDetach.VolumeName)
+			if markDeviceUnmountedErr != nil {
+				// On failure, return error. Caller will log and retry.
+				return deviceToDetach.GenerateErrorDetailed("MarkDeviceAsUnmounted failed", markDeviceUnmountedErr)
+			}
+			return nil
+		} else if err == nil {
+			glog.Infof(deviceToDetach.GenerateMsgDetailed("UnmountDevice succeeded", ""))
+			// Update actual state of world
+			markDeviceUnmountedErr := actualStateOfWorld.MarkDeviceAsUnmounted(
+				deviceToDetach.VolumeName)
+			if markDeviceUnmountedErr != nil {
+				// On failure, return error. Caller will log and retry.
+				return deviceToDetach.GenerateErrorDetailed("MarkDeviceAsUnmounted failed", markDeviceUnmountedErr)
+			}
+			glog.Infof("#### DEBUG LOG ####: GenerateUnmountDeviceFunc unmapDevice path  deviceMountPath: %s", deviceMountPath)
+			return nil
+		}
+		return deviceToDetach.GenerateErrorDetailed(
+			"UnmountDevice failed",
+			err)
+	}, attachableVolumePlugin.GetPluginName(), nil
+}
+
+func (og *operationGenerator) GenerateMapVolumeFunc(
+	waitForAttachTimeout time.Duration,
+	volumeToMount VolumeToMount,
+	actualStateOfWorld ActualStateOfWorldMounterUpdater,
+	mounter mount.Interface) (func() error, string, error) {
+
+	var blockVolumeMapper volume.BlockVolumeMapper
+	pv := volumeToMount.VolumeSpec.PersistentVolume
+	if pv == nil {
+		return nil, "", volumeToMount.GenerateErrorDetailed("MapVolume. Find PersistentVolume failed. PersistentVolume is nil.", nil)
+	}
+	claimName := pv.Spec.ClaimRef.Name
+	pvc, _ := og.kubeClient.Core().PersistentVolumeClaims(volumeToMount.Pod.ObjectMeta.Namespace).Get(claimName, metav1.GetOptions{})
```

Comment:
Can you use the volume mode from the PV instead of making an API call to get the PVC?

Reply:
Ack.
Based on the latest bind matrix, we can decide VolumeMode by only referring the PV volumeMode after passing the pv-binder.